File: go/internal/csvdiff/rowstore.go
package csvdiff

import (
	"cmp"
	"slices"
)

// RowStore is the in-memory half of a comparison: it owns the parts that must
// not drift between engines — first-occurrence-wins de-duplication, the
// duplicate-key report, the full outer join and the sparse cell diffs — so an
// engine only has to supply normalised rows.
//
// The DuckDB engine does all of this in SQL instead and does not use this type.
type RowStore struct {
	opt      Options
	key      []string
	compared []string
	keySize  int

	order       []string // key order of first appearance, so the join is deterministic
	first       map[string][]*string
	occurrences map[string]int
	rows        int64
}

// NewRowStore prepares a store for rows shaped as key columns then compared columns.
func NewRowStore(opt Options, compared []string) *RowStore {
	return &RowStore{
		opt:         opt,
		key:         opt.Key,
		compared:    compared,
		keySize:     len(opt.Key),
		first:       make(map[string][]*string),
		occurrences: make(map[string]int),
	}
}

// Add records one already-normalised row.
func (s *RowStore) Add(row []*string) {
	s.rows++
	k := KeyOf(row, s.keySize)
	s.occurrences[k]++
	if _, seen := s.first[k]; !seen {
		s.first[k] = row
		s.order = append(s.order, k)
	}
}

// Rows is the number of rows read, duplicates included.
func (s *RowStore) Rows() int64 { return s.rows }

// UniqueKeys is the number of distinct composite keys.
func (s *RowStore) UniqueKeys() int64 { return int64(len(s.occurrences)) }

// DuplicateKeys is how many keys appear more than once.
func (s *RowStore) DuplicateKeys() int64 {
	var n int64
	for _, c := range s.occurrences {
		if c > 1 {
			n++
		}
	}
	return n
}

// DuplicateRows is how many rows carry a duplicated key.
func (s *RowStore) DuplicateRows() int64 {
	var n int64
	for _, c := range s.occurrences {
		if c > 1 {
			n += int64(c)
		}
	}
	return n
}

// DuplicateSection is the duplicate-key list: most duplicated first, then by key.
func (s *RowStore) DuplicateSection() Section {
	type entry struct {
		row []*string
		n   int
	}
	var entries []entry
	for _, k := range s.order {
		if c := s.occurrences[k]; c > 1 {
			entries = append(entries, entry{s.first[k], c})
		}
	}
	slices.SortStableFunc(entries, func(x, y entry) int {
		if c := cmp.Compare(y.n, x.n); c != 0 {
			return c
		}
		return CompareKeys(x.row, y.row, s.keySize)
	})

	limit := min(len(entries), s.opt.MaxRows)
	rows := make([][]any, 0, limit)
	for _, e := range entries[:limit] {
		row := make([]any, 0, s.keySize+1)
		for i := range s.keySize {
			row = append(row, e.row[i])
		}
		rows = append(rows, append(row, e.n))
	}
	return Section{
		Cols:      append(append([]string{}, s.key...), "count"),
		Rows:      rows,
		Truncated: len(entries) > s.opt.MaxRows,
	}
}

// Joined is everything the full outer join produces, before the sections are capped.
type Joined struct {
	Counts   Counts
	Columns  []ColumnStat
	Changed  [][]any
	Added    [][]any
	Removed  [][]any
	ChangedA [][]*string // A-side rows aligned with Changed, for --export-dir
	ChangedB [][]*string
}

// Join is the full outer join of two stores on the composite key.
func Join(a, b *RowStore, opt Options) Joined {
	keySize := len(opt.Key)
	nc := len(a.compared)

	changedPer := make([]int64, nc)
	blankedPer := make([]int64, nc)
	filledPer := make([]int64, nc)

	var changed, added, removed [][]any
	var changedA, changedB [][]*string
	var matched int64

	// Iterate in first-appearance order so a run is reproducible.
	for _, k := range a.order {
		ar := a.first[k]
		br, ok := b.first[k]
		if !ok {
			removed = append(removed, toAny(ar))
			continue
		}
		matched++
		var cells []CellDiff
		for i := range nc {
			x, y := ar[keySize+i], br[keySize+i]
			if Differs(x, y, opt) {
				cells = append(cells, CellDiff{Column: i, A: x, B: y})
				changedPer[i]++
				if y == nil {
					blankedPer[i]++
				}
				if x == nil {
					filledPer[i]++
				}
			}
		}
		if len(cells) > 0 {
			row := make([]any, 0, keySize+1)
			for i := range keySize {
				row = append(row, ar[i])
			}
			changed = append(changed, append(row, cells))
			changedA = append(changedA, ar)
			changedB = append(changedB, br)
		}
	}
	for _, k := range b.order {
		if _, ok := a.first[k]; !ok {
			added = append(added, toAny(b.first[k]))
		}
	}

	columns := make([]ColumnStat, nc)
	for i := range nc {
		columns[i] = ColumnStat{Name: a.compared[i], Changed: changedPer[i], Blanked: blankedPer[i], Filled: filledPer[i]}
	}

	sortRowsByKey(added, keySize)
	sortRowsByKey(removed, keySize)
	sortChangedTogether(changed, changedA, changedB, keySize)

	counts := Counts{
		ARows: a.Rows(), BRows: b.Rows(),
		AKeys: a.UniqueKeys(), BKeys: b.UniqueKeys(),
		Matched: matched, Unchanged: matched - int64(len(changed)), Changed: int64(len(changed)),
		Added: int64(len(added)), Removed: int64(len(removed)),
		ADupKeys: a.DuplicateKeys(), ADupRows: a.DuplicateRows(),
		BDupKeys: b.DuplicateKeys(), BDupRows: b.DuplicateRows(),
	}
	return Joined{counts, columns, changed, added, removed, changedA, changedB}
}

func toAny(row []*string) []any {
	out := make([]any, len(row))
	for i, v := range row {
		out[i] = v
	}
	return out
}

func sortRowsByKey(rows [][]any, keySize int) {
	slices.SortStableFunc(rows, func(x, y []any) int {
		for i := range keySize {
			a, _ := x[i].(*string)
			b, _ := y[i].(*string)
			if a == nil && b == nil {
				continue
			}
			if a == nil {
				return 1
			}
			if b == nil {
				return -1
			}
			if c := cmp.Compare(*a, *b); c != 0 {
				return c
			}
		}
		return 0
	})
}

// sortChangedTogether sorts changed rows by key while keeping the parallel A and
// B row slices in step, which --export-dir relies on.
func sortChangedTogether(changed [][]any, changedA, changedB [][]*string, keySize int) {
	idx := make([]int, len(changed))
	for i := range idx {
		idx[i] = i
	}
	slices.SortStableFunc(idx, func(p, q int) int {
		return CompareKeys(changedA[p], changedA[q], keySize)
	})

	c := make([][]any, len(changed))
	ca := make([][]*string, len(changed))
	cb := make([][]*string, len(changed))
	for n, i := range idx {
		c[n], ca[n], cb[n] = changed[i], changedA[i], changedB[i]
	}
	copy(changed, c)
	copy(changedA, ca)
	copy(changedB, cb)
}

// CapSection caps a row list at --max-rows and records whether anything was cut.
func CapSection(cols []string, rows [][]any, maxRows int) Section {
	if rows == nil {
		rows = [][]any{}
	}
	truncated := len(rows) > maxRows
	return Section{Cols: cols, Rows: rows[:min(len(rows), maxRows)], Truncated: truncated}
}

File: rust/src/rowstore.rs
//! The in-memory half of a comparison.
//!
//! [`RowStore`] owns the parts that must not drift between engines —
//! first-occurrence-wins de-duplication, the duplicate-key report, the full
//! outer join and the sparse cell diffs — so an engine only has to supply
//! normalised rows.
//!
//! The DuckDB engine does all of this in SQL instead and does not use this type.

use std::collections::HashMap;

use crate::columns::{compare_keys, differs, key_of};
use crate::contract::{Cell, CellDiff, ColumnStat, Counts, Section, Val};
use crate::options::Options;

/// Rows shaped as key columns followed by compared columns.
pub struct RowStore {
    key: Vec<String>,
    compared: Vec<String>,
    key_size: usize,
    max_rows: usize,

    /// Key order of first appearance, so the join is deterministic.
    order: Vec<String>,
    first: HashMap<String, Vec<Val>>,
    occurrences: HashMap<String, u32>,
    rows: i64,
}

impl RowStore {
    pub fn new(opt: &Options, compared: &[String]) -> Self {
        RowStore {
            key: opt.key.clone(),
            compared: compared.to_vec(),
            key_size: opt.key.len(),
            max_rows: opt.max_rows,
            order: Vec::new(),
            first: HashMap::new(),
            occurrences: HashMap::new(),
            rows: 0,
        }
    }

    /// Records one already-normalised row.
    pub fn add(&mut self, row: Vec<Val>) {
        self.rows += 1;
        let k = key_of(&row, self.key_size);
        *self.occurrences.entry(k.clone()).or_insert(0) += 1;
        if !self.first.contains_key(&k) {
            self.order.push(k.clone());
            self.first.insert(k, row);
        }
    }

    /// The number of rows read, duplicates included.
    pub fn rows(&self) -> i64 {
        self.rows
    }

    /// The number of distinct composite keys.
    pub fn unique_keys(&self) -> i64 {
        self.occurrences.len() as i64
    }

    /// How many keys appear more than once.
    pub fn duplicate_keys(&self) -> i64 {
        self.occurrences.values().filter(|&&c| c > 1).count() as i64
    }

    /// How many rows carry a duplicated key.
    pub fn duplicate_rows(&self) -> i64 {
        self.occurrences
            .values()
            .filter(|&&c| c > 1)
            .map(|&c| i64::from(c))
            .sum()
    }

    /// The duplicate-key list: most duplicated first, then by key.
    pub fn duplicate_section(&self) -> Section {
        let mut entries: Vec<(&Vec<Val>, u32)> = self
            .order
            .iter()
            .filter_map(|k| {
                let count = self.occurrences[k];
                (count > 1).then(|| (&self.first[k], count))
            })
            .collect();
        entries
            .sort_by(|(x, xn), (y, yn)| yn.cmp(xn).then_with(|| compare_keys(x, y, self.key_size)));

        let truncated = entries.len() > self.max_rows;
        entries.truncate(self.max_rows);
        let rows = entries
            .into_iter()
            .map(|(row, count)| {
                let mut out: Vec<Cell> = row[..self.key_size]
                    .iter()
                    .map(|v| Cell::Value(v.clone()))
                    .collect();
                out.push(Cell::Count(i64::from(count)));
                out
            })
            .collect();

        let mut cols = self.key.clone();
        cols.push("count".to_string());
        Section {
            cols,
            rows,
            truncated,
        }
    }

    /// The compared column names this store was built for.
    pub fn compared(&self) -> &[String] {
        &self.compared
    }
}

/// Everything the full outer join produces, before the sections are capped.
pub struct Joined {
    pub counts: Counts,
    pub columns: Vec<ColumnStat>,
    pub changed: Vec<Vec<Cell>>,
    pub added: Vec<Vec<Cell>>,
    pub removed: Vec<Vec<Cell>>,
    /// A-side rows aligned with `changed`, for `--export-dir`.
    pub changed_a: Vec<Vec<Val>>,
    pub changed_b: Vec<Vec<Val>>,
}

/// The full outer join of two stores on the composite key.
pub fn join(a: &RowStore, b: &RowStore, opt: &Options) -> Joined {
    let key_size = opt.key.len();
    let nc = a.compared.len();

    let mut changed_per = vec![0i64; nc];
    let mut blanked_per = vec![0i64; nc];
    let mut filled_per = vec![0i64; nc];

    let mut changed: Vec<Vec<Cell>> = Vec::new();
    let mut changed_a: Vec<Vec<Val>> = Vec::new();
    let mut changed_b: Vec<Vec<Val>> = Vec::new();
    let mut added: Vec<Vec<Cell>> = Vec::new();
    let mut removed: Vec<Vec<Cell>> = Vec::new();
    let mut matched = 0i64;

    // Iterate in first-appearance order so a run is reproducible.
    for k in &a.order {
        let ar = &a.first[k];
        let Some(br) = b.first.get(k) else {
            removed.push(to_cells(ar));
            continue;
        };
        matched += 1;

        let mut cells: Vec<CellDiff> = Vec::new();
        for i in 0..nc {
            let (x, y) = (&ar[key_size + i], &br[key_size + i]);
            if differs(x, y, opt) {
                cells.push(CellDiff {
                    column: i,
                    a: x.clone(),
                    b: y.clone(),
                });
                changed_per[i] += 1;
                if y.is_none() {
                    blanked_per[i] += 1;
                }
                if x.is_none() {
                    filled_per[i] += 1;
                }
            }
        }
        if !cells.is_empty() {
            let mut row: Vec<Cell> = ar[..key_size]
                .iter()
                .map(|v| Cell::Value(v.clone()))
                .collect();
            row.push(Cell::Diffs(cells));
            changed.push(row);
            changed_a.push(ar.clone());
            changed_b.push(br.clone());
        }
    }
    for k in &b.order {
        if !a.first.contains_key(k) {
            added.push(to_cells(&b.first[k]));
        }
    }

    let columns = (0..nc)
        .map(|i| ColumnStat {
            name: a.compared[i].clone(),
            changed: changed_per[i],
            blanked: blanked_per[i],
            filled: filled_per[i],
        })
        .collect();

    sort_by_key_cells(&mut added, key_size);
    sort_by_key_cells(&mut removed, key_size);
    sort_changed_together(&mut changed, &mut changed_a, &mut changed_b, key_size);

    let counts = Counts {
        a_rows: a.rows(),
        b_rows: b.rows(),
        a_keys: a.unique_keys(),
        b_keys: b.unique_keys(),
        matched,
        unchanged: matched - changed.len() as i64,
        changed: changed.len() as i64,
        added: added.len() as i64,
        removed: removed.len() as i64,
        a_dup_keys: a.duplicate_keys(),
        a_dup_rows: a.duplicate_rows(),
        b_dup_keys: b.duplicate_keys(),
        b_dup_rows: b.duplicate_rows(),
    };

    Joined {
        counts,
        columns,
        changed,
        added,
        removed,
        changed_a,
        changed_b,
    }
}

fn to_cells(row: &[Val]) -> Vec<Cell> {
    row.iter().map(|v| Cell::Value(v.clone())).collect()
}

fn sort_by_key_cells(rows: &mut [Vec<Cell>], key_size: usize) {
    rows.sort_by(|x, y| {
        for i in 0..key_size {
            let ord = match (x[i].value(), y[i].value()) {
                (None, None) => std::cmp::Ordering::Equal,
                (None, Some(_)) => std::cmp::Ordering::Greater,
                (Some(_), None) => std::cmp::Ordering::Less,
                (Some(a), Some(b)) => a.cmp(b),
            };
            if ord != std::cmp::Ordering::Equal {
                return ord;
            }
        }
        std::cmp::Ordering::Equal
    });
}

/// Sorts changed rows by key while keeping the parallel A and B row lists in
/// step, which `--export-dir` relies on.
fn sort_changed_together(
    changed: &mut Vec<Vec<Cell>>,
    changed_a: &mut Vec<Vec<Val>>,
    changed_b: &mut Vec<Vec<Val>>,
    key_size: usize,
) {
    let mut idx: Vec<usize> = (0..changed.len()).collect();
    idx.sort_by(|&p, &q| compare_keys(&changed_a[p], &changed_a[q], key_size));

    *changed = reorder(std::mem::take(changed), &idx);
    *changed_a = reorder(std::mem::take(changed_a), &idx);
    *changed_b = reorder(std::mem::take(changed_b), &idx);
}

fn reorder<T>(mut items: Vec<T>, idx: &[usize]) -> Vec<T> {
    let mut slots: Vec<Option<T>> = items.drain(..).map(Some).collect();
    idx.iter()
        .map(|&i| slots[i].take().expect("each index used once"))
        .collect()
}
